### PR TITLE
[WIP] Code implementation of Conv-LoRA

### PR DESCRIPTION
**Pull request by Harry-zzh**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Harry-zzh** [comment]
> Great work! One real quick question, have you tried conv-lora on standard text tasks instead of image/SAM tasks? If so, how was it? If havnt tried, do you think it is a more general-purpose PEFT method, or it is more of a SAM/CV-specific approach? Thanks a lot!

Thank you for your question. While I haven't tried text tasks yet, my understanding is that Conv-LoRA is primarily designed for image tasks.

Conv-LoRA incorporates local priors into image features at appropriate scales, considering potential variations in object scale. This involves interpolating image features to larger scales than default and subsequently employing convolution operations for injecting local priors. In our paper, we find that interpolating features to larger scales for local prior injection is more beneficial, given that features in the Vision Transformer (ViT) are downscaled by a factor (e.g., 16) from the original continuous image.

However, texts are 1-D discrete sequences and lack a concept akin to "object scale". Consequently, considering the feature processing in Conv-LoRA and its motivation, it is unsuitable for text tasks.

File: multimodal/src/autogluon/multimodal/models/conv_lora/moe_conv.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
+from typing import Any, Callable, Dict, Iterator, List, Mapping, Optional, Set, Tuple, TypeVar, Union, overload
+
+import numpy as np
+import torch
+from torch import nn
+from torch.distributions.normal import Normal
+
+
+class MoEConv(nn.Module):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/conv_lora/moe_conv.py:9]: Need a more accurate name? `MoEGate`? This class doesn't contain convolutions and is to determine gates.

File: multimodal/src/autogluon/multimodal/models/conv_lora/modeling_sam.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/conv_lora/modeling_sam.py:1]: How about creating a folder `custom_hf_models` and put this file there with a new name `modeling_sam_for_conv_lora`? The consideration is that we may encounter similar situations of hosting customized hf model definitions in the future, so all these kinds of models can go here.

File: multimodal/src/autogluon/multimodal/models/conv_lora/adaptation_layers.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/conv_lora/adaptation_layers.py:1]: Since we already have a `multimodal/src/autogluon/multimodal/models/adaptation_layers.py`, how about we put the definitions of `ConvLoRALinear`, `MoEConv`, and `SparseDispatcher` in that file?

File: multimodal/src/autogluon/multimodal/configs/optimization/default.yaml [hunk 2]
```diff
@@ -34,6 +34,7 @@ optimization:
       - "^o$"
     r: 8
     alpha: 8
+    conv_lora_expert_num: 8 # default setting for Conv-LoRA
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/optimization/default.yaml:37]: Any other important hyperparameters for conv-lora?

File: multimodal/src/autogluon/multimodal/constants.py [hunk 3]
```diff
@@ -139,6 +139,7 @@
 MAE = "mae"
 BER = "ber"
 IOU = "iou"
+MOE_LOSS = "moe_loss"
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/constants.py:142]: We can add a new section `# loss` and put `MOE_LOSS = "moe_loss"` there.

File: multimodal/src/autogluon/multimodal/models/conv_lora/adaptation_layers.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        r: int = 0,
+        lora_alpha: int = 1,
+        lora_dropout: float = 0.0,
+        fan_in_fan_out: bool = False,  # Set this to True if the layer to replace stores weight like (fan_in, fan_out)
+        merge_weights: bool = False,
```
Comment (2catycm) [inline multimodal/src/autogluon/multimodal/models/conv_lora/adaptation_layers.py:50]: Is Conv-Lora reparameterizable? 
It is more complicated than lora, lora just merge the weights by multiplying the matrices.
but here we have convolutions.

Comment (Harry-zzh) [reply multimodal/src/autogluon/multimodal/models/conv_lora/adaptation_layers.py:50]: Conv-LoRA is not reparameterizable mainly due to its interpolation operation. 
Actually convolutions is not the main reason because convolution layer could be re-parameterized into FC layer in some cases. You could refer to papers about structural re-parameterization for more details.

File: multimodal/src/autogluon/multimodal/models/sam.py [hunk 5]
```diff
@@ -359,6 +369,15 @@ def mask_label_key(self):
...
 
+    def train(self, mode: bool = True):
+        super().train(mode)
+        for module in self.modules():
+            if isinstance(module, ConvLoRALinear):
+                self.output_moe_loss = True
+                return self
+
+        return self
+
```
[the remark this answers is not in the document]
Comment (Harry-zzh) [reply multimodal/src/autogluon/multimodal/models/sam.py:380]: We need the MoE loss when calculating the validation loss. During the validation process, the module mode is set to "eval". So we cannot distinguish the validation and inference process here.

**Review by zhiqiangdon — APPROVED** [review]
LGTM!
